### ISSUE TITLE
JDK-8207942: Add new protected VirtualFlow methods for subclassing

### ISSUE THREAD

**Issue by eugener** [javafxports/openjdk-jfx#151]
Related JBS issues:
https://bugs.openjdk.java.net/browse/JDK-8207942
https://bugs.openjdk.java.net/browse/JDK-8207957

From JDK 9 onward it is not possible to subclass VirtualFlow and many libraries/applications cannot produce adequate releases. This can finally be fixed in OpenJFX 12.

**Comment by Maxoudela** [comment]
I can provide a pull request if you like.

**Comment by kevinrushforth** [comment]
What we need first is a discussion on the API. A [WIP] pull request might be a starting point for that discussion, but since this will be new feature with new public API, it will need a [CSR](https://wiki.openjdk.java.net/display/csr/Main). I think @eugener plans to do this, so please wait for him to weigh in on this.

**Comment by eugener** [comment]
In my view there are two approaches possible: 

1. Make some of the existing methods protected to allow for direct extension of `VirtualFlow` as Samir suggests
2. Change the `VirtualFlow` to use a pluggable implementations, i.e define APIs based on interfaces which can be plugged into `VirtualFlow`, so it can call them when required. 

Samir (@Maxoudela) is the one who filed the ttps://bugs.openjdk.java.net/browse/JDK-8207942, So I would be ok if we start the discussion with his PR, provided that approach 1 is deemed to be more favorable.

**Comment by kevinrushforth** [comment]
Approach 1 might be be a quicker path, and it might be nice to see what such an API would look like, so if Samir wants to generate a PR as a discussion point, that seems fine.

@eugener - are you interested in exploring option 2? If so, that might be better done over email rather than a prototype, but even that might possibly be done as a [WIP] PR. One question though: you say "change" VirtualFlow. You mean compatibly, I hope?

**Comment by eugener** [comment]
@kevinrushforth As much as it is possible it should be fully compatible. The internals of it will simply call the interface methods. This is all hypothetical of course. reality might turn out to be different :)

I will try to experiment with it, as long as you think it might be somewhat viable approach.

**Comment by kevinrushforth** [comment]
@eugener If you think there might be value in it, and if the resulting API would be something that would "fit" with the existing API, then you might spend a little time prototyping it to see what it would look like.

**Comment by Maxoudela** [comment]
I have not forgotten you guys. I'll make the PR as soon as possible as a discussion point. 
I just need to install and build OpenJFX, and also make the PR the right way. 
As a side note, where can I share some thoughts about https://wiki.openjdk.java.net/display/OpenJFX/Building+OpenJFX ? I may have some comments that can help others.

I've been trying to build OpenJFX on my computer all afternoon, bumping into every possible issues. Right now I'm stuck on :graphics:ccWinGlass with that exception :

```

* What went wrong:
Execution failed for task ':graphics:ccWinGlass'.
> java.util.concurrent.ExecutionException: org.gradle.process.internal.ExecException: Process 'command 'C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.15.26726/bin/HostX64/x64/cl.exe'' finished with non-zero exit value 2

* Try:
Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Exception is:
org.gradle.api.tasks.TaskExecutionException: Execution failed for task ':graphics:ccWinGlass'.
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:110)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.OutputDirectoryCreatingTaskExecuter.execute(OutputDirectoryCreatingTaskExecuter.java:51)
        at org.gradle.api.internal.tasks.execution.SkipUpToDateTaskExecuter.execute(SkipUpToDateTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.ResolveTaskOutputCachingStateExecuter.execute(ResolveTaskOutputCachingStateExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ValidatingTaskExecuter.execute(ValidatingTaskExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipEmptySourceFilesTaskExecuter.execute(SkipEmptySourceFilesTaskExecuter.java:101)
        at org.gradle.api.internal.tasks.execution.FinalizeInputFilePropertiesTaskExecuter.execute(FinalizeInputFilePropertiesTaskExecuter.java:44)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:91)
        at org.gradle.api.internal.tasks.execution.ResolveTaskArtifactStateTaskExecuter.execute(ResolveTaskArtifactStateTaskExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:59)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:54)
        at org.gradle.api.internal.tasks.execution.ExecuteAtMostOnceTaskExecuter.execute(ExecuteAtMostOnceTaskExecuter.java:43)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:34)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.run(EventFiringTaskExecuter.java:51)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:46)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$ExecuteTaskAction.execute(DefaultTaskExecutionGraph.java:262)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$ExecuteTaskAction.execute(DefaultTaskExecutionGraph.java:246)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:136)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker$1.execute(DefaultTaskPlanExecutor.java:130)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.execute(DefaultTaskPlanExecutor.java:201)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.executeWithTask(DefaultTaskPlanExecutor.java:192)
        at org.gradle.execution.taskgraph.DefaultTaskPlanExecutor$TaskExecutorWorker.run(DefaultTaskPlanExecutor.java:130)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
Caused by: org.gradle.internal.UncheckedException: java.util.concurrent.ExecutionException: org.gradle.process.internal.ExecException: Process 'command 'C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.15.26726/bin/HostX64/x64/cl.exe'' finished with non-zero exit value 2
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:63)
        at org.gradle.internal.UncheckedException.throwAsUncheckedException(UncheckedException.java:40)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:76)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:46)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:39)
        at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:26)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:794)
        at org.gradle.api.internal.AbstractTask$TaskActionWrapper.execute(AbstractTask.java:761)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter$1.run(ExecuteActionsTaskExecuter.java:131)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:317)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:309)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:185)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:97)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeAction(ExecuteActionsTaskExecuter.java:120)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeActions(ExecuteActionsTaskExecuter.java:99)
        ... 30 more
Caused by: java.util.concurrent.ExecutionException: org.gradle.process.internal.ExecException: Process 'command 'C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.15.26726/bin/HostX64/x64/cl.exe'' finished with non-zero exit value 2
        at java_util_concurrent_Future$get.call(Unknown Source)
        at NativeCompileTask$_compile_closure6.doCall(NativeCompileTask.groovy:161)
        at jdk.internal.reflect.GeneratedMethodAccessor129.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at NativeCompileTask.compile(NativeCompileTask.groovy:161)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:73)
        ... 43 more
Caused by: org.gradle.process.internal.ExecException: Process 'command 'C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14.15.26726/bin/HostX64/x64/cl.exe'' finished with non-zero exit value 2
        at org.gradle.process.internal.DefaultExecHandle$ExecResultImpl.assertNormalExitValue(DefaultExecHandle.java:395)
        at org.gradle.process.internal.DefaultExecAction.execute(DefaultExecAction.java:37)
        at org.gradle.api.internal.file.DefaultFileOperations.exec(DefaultFileOperations.java:198)
        at org.gradle.api.internal.project.DefaultProject.exec(DefaultProject.java:1097)
        at org.gradle.api.internal.project.DefaultProject.exec(DefaultProject.java:1092)
        at org.gradle.api.Project$exec$8.call(Unknown Source)
        at CCTask.doCompile(CCTask.groovy:52)
        at jdk.internal.reflect.GeneratedMethodAccessor152.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at NativeCompileTask.this$dist$invoke$3(NativeCompileTask.groovy)
        at NativeCompileTask$1.methodMissing(NativeCompileTask.groovy)
        at jdk.internal.reflect.GeneratedMethodAccessor122.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at NativeCompileTask$1.run(NativeCompileTask.groovy:150)
```
Regarding our matter, maybe building and testing against "controls" is enough?

**Comment by kevinrushforth** [comment]
> As a side note, where can I share some thoughts about https://wiki.openjdk.java.net/display/OpenJFX/Building+OpenJFX ? I may have some comments that can help others.

Maybe you can file an issue in this GitHub project if you have specific suggestions.

As for your build failures, it's hard to say, but I recommend the following steps to try again with a clean build:

1. Stop any gradle daemons that might be running with `gradle --stop` (and disable them if you haven't already)

2. Manually delete the build directory: `rm -rf build`

3. Run `gradle clean`

4. Remove the following:

    `rm -rf .gradle buildSrc/build buildSrc/.gradle`

The try a build again. If it fails, look at the contents of the following file and see if anything is amiss (either undefined or pointing to a nonexistent path) -- `build/windows_tools.properties`

**Comment by Maxoudela** [comment]
I followed your steps but that did not solve the issue. 
I have everything accessible in `build/windows_tools.properties`.
I gave up and finally built the whole thing on my Mac in 3m51s. Whereas the graphic compilation and cleanup on Windows takes up to 20 minutes.. Don't know what is the issue.

So I'll try to develop on my Mac and make the PR.

**Comment by kevinrushforth** [comment]
> I followed your steps but that did not solve the issue.

It's possible you have discovered a bug relating to compiling with a newer version of Visual Studio. At some point we should figure out what the problem is and how to solve it.

**Comment by Maxoudela** [comment]
Indeed.. I'll maybe try to run the build using other options in order to see why VisualStudio is actually failing. Maybe activate some logs on Visual Studio? 

Anyway, I believed I've made the pull request. Don't hesitate to tell me if I've made something wrong regarding the commit or the branches.

I'll try to use my own version of OpenJFX12 against the SpreadsheetView in order to see if the API change is sufficient (or too wide). I currently have some difficulty building a project using JDK-11. My plan is to follow this tutorial : http://docs.gluonhq.com/javafx11/#maven and manually replace the downloaded OpenJFX jar with my own. 
If you have any better alternative or advice, I'll gladly take them.

As a side-note, I'll be on holiday for one week so I'll try to be around but there's a possibility that I remain completely afk during that time.

**Comment by kevinrushforth** [comment]
I'll take a preliminary look at the PR early next week most likely.

**Comment by eugener** [comment]
I spent several nights looking into pluggable approach to `VirtualFlow`. IMO it does not have as much benefits as approach 1. 

- Minimal benefit can be achieved by extracting VirtualFlow as an interface. Actual implementation is pretty complex and I don't think many people would venture into a full re-implementation.
- I also looked into implementation of external plugin API. But this approach prohibits access to internal state of current implementation, which is important for many developers

Unless there are other ideas, I think Samir's approach of making key methods protected is currently our best bet.
